src/parser/mod.rs

```diff
@@ -2800,11 +2814,40 @@ impl<'a> Parser<'a> {
                 format: None,
             })
         } else if Token::ExclamationMark == tok {
-            // PostgreSQL factorial operation
-            Ok(Expr::UnaryOp {
-                op: UnaryOperator::PGPostfixFactorial,
-                expr: Box::new(expr),
-            })
+            if self.dialect.supports_factorial_operator() {
+                match expr {
+                    Expr::Value(_) | Expr::Identifier(_) | Expr::Nested(_) | Expr::BinaryOp{..} => Ok(Expr::UnaryOp {
+                        op: UnaryOperator::PGPostfixFactorial,
+                        expr: Box::new(expr),
+                    }),
+                    _ => {
+                        self.expected(
+                            "Value or Identifier or  Nested or BinaryOp struct before factorial operator(!)", self.peek_token())
```

Comment:
Because both `Hive` and `PostgreSQL` dialects support the exclamation mark, we check the token before the exclamation mark to determine if it's a factorial operator, while also avoiding the incorrect identification of Hive `BangNot` syntax as PostgreSQL  `PostfixFactorial`.

Reply:
Yes, you're right. After testing, returning `Ok(Expr::UnaryOp ...)` directly here doesn't affect the result.